gammapy/background/ring.py

```diff
@@ -2,22 +2,236 @@
 """Ring background estimation.
 """
 from __future__ import absolute_import, division, print_function, unicode_literals
+from itertools import product
 import numpy as np
-from astropy.convolution import Ring2DKernel
+from astropy.convolution import Ring2DKernel, Tophat2DKernel
+import astropy.units as u
 from ..image import SkyImageList, SkyImage
+from ..image.utils import scale_cube_fft
 
 
 __all__ = [
+    'AdaptiveRingBackgroundEstimator',
     'RingBackgroundEstimator',
     'ring_r_out',
     'ring_area_factor',
     'ring_alpha',
 ]
 
 
-class RingBackgroundEstimator(object):
+class AdaptiveRingBackgroundEstimator(object):
+    """
+    Adaptive ring background algorithm.
+
+    This alogrithm extends the standard `RingBackground` method by adpating the
+    size of the ring to achieve a minimum on / off exposure ratio (alpha) in regions
+    where the area to estimate the background from is limited.
+
+    Here is an illustration of the method:
+
+    Parameters
+    ----------
+    r_in : `~astropy.units.Quantity`
+        Inner radius of the ring.
+    r_out_max : `~astropy.units.Quantity`
+        Maximal outer radius of the ring.
+    width : `~astropy.units.Quantity`
+        Width of the ring.
+    stepsize : `~astropy.units.Quantity` (0.02 deg)
+        Stepsize used for increasing the radius.
+    threshold : float (0.1)
+        Threshold above which the adaptive ring takes action.
+    theta : `~astropy.units.Quantity` (0.22 deg)
+        Integration radius used for alpha computation.
+    method : {'const. width', 'const. r_in'}
+        Adaptive ring method.
+
+    Examples
+    --------
+    Here's an example how to use the `AdaptiveRingBackgroundEstimator`:
+
+    .. code::
+
+        from astropy import units as u
+        from gammapy.background import AdaptiveRingBackgroundEstimator
+        from gammapy.image import SkyImageList
+
+        filename = '$GAMMAPY_EXTRA/test_datasets/unbundled/poisson_stats_image/input_all.fits.gz'
+        images = SkyImageList.read(filename)
+        images['exposure'].name = 'exposure_on'
+        adaptive_ring_bkg = RingBackgroundEstimator(r_in=0.22 * u.deg,
+                                                    r_out_max=0.8 * u.deg,
+                                                    width=0.1 * u.deg)
+        results = adaptive_ring_bkg.run(images)
+        results['background'].show()
+
+    See Also
+    --------
+    RingBackgroundEstimator, KernelBackgroundEstimator
+
     """
-    Ring background method for cartesian coordinates.
+    def __init__(self, r_in, r_out_max, width, stepsize=None,
+                 threshold=0.1, theta=None, method='const. width'):
+
+        # default values
```

Comment:
I've thought about this a lot and had discussions with others.
The plan is to use Quantity objects as default kwargs, even if they are mutable.
So I'd be +1 to put the default values in the signature and remove them from here and the docstring, to make maintenance a little easier (only have the default in one place, not two).